src/js/layout.js

```diff
@@ -110,53 +110,167 @@ class Layout {
-			if (window.innerHeight + window.pageYOffset >= document.body.scrollHeight) {
-				return;
+	/**
+	 * Add aria-current to the correspoding link in the navigation for the hash in the url if one exists
+	 * @private
+	 * @returns {void}
+	 */
+	highlightNavigationFromLocation() {
+		if (location.hash) {
+			// on page load, highlight the nav item that corresponds to the url
+			this.navAnchors.forEach((anchor, index) => {
+				const currentLocation = anchor.hash === location.hash;
+				const defaultLocation = location.hash === '' && index === 0;
+				if (currentLocation || defaultLocation) {
+					anchor.setAttribute('aria-current', 'location');
+					this.highlightedHeadingIndex = index;
+				} else {
+					anchor.setAttribute('aria-current', 'false');
+				}
+			});
+		}
+	}
+
+	/**
+	 * Add aria-current to the correspoding link in the navigation for the header that we think
+	 * should be highlighted based on scroll position
+	 * @private
+	 * @returns {void}
+	 */
+	 setupIntersectionObserversForHeadings() {
+		function getY(domRect) {
+			return Object.prototype.hasOwnProperty.call(domRect, 'y')
+				? domRect.y
+				: domRect.top;
+		}
+
+		const observer = new IntersectionObserver(
+			entries => {
+				let headingIndexToHighlight = this.highlightedHeadingIndex;
+
+				// Record index of which headings are above or below the intersection target
+				const above = [];
+				const below = [];
+				entries.forEach(entry => {
+					const intersectingElemIdx = this.navHeadings.findIndex(
+						navheading => navheading === entry.target
+					);
+					const isAbove =
+						getY(entry.boundingClientRect) < (getY(entry.rootBounds || {}) || 0);
+					if (isAbove) {
+						above.push(intersectingElemIdx);
+					} else {
+						below.push(intersectingElemIdx);
+					}
+				});
+
+				// If there are headings above the intersection target,
+				// set the active heading as the last one which is above the intersection target
+				if (above.length > 0) {
+					// Find the last heading index which is above the intersection target
+					headingIndexToHighlight = Math.max(...above);
+				} else if (below.length > 0) {
+					// Find the first heading index which is below the intersection target
+					const minIndex = Math.min(...below);
+					if (minIndex <= this.highlightedHeadingIndex) {
+						// If there are no headings above  the intersection target and the current
+						// active heading is later down the page then use the first heading which is below
+						headingIndexToHighlight = minIndex - 1 >= 0 ? minIndex - 1 : 0;
+					}
+				}
+				this.navAnchors.forEach((anchor, index) => {
+					if (headingIndexToHighlight === index) {
+						anchor.setAttribute('aria-current', 'location');
+					} else {
+						anchor.setAttribute('aria-current', 'false');
+					}
+				});
+				this.highlightedHeadingIndex = headingIndexToHighlight;
+			}, {
+				rootMargin: `-${
+					window.getComputedStyle(document.querySelector('h2')).fontSize
```

Comment:
Todo: Something better here such as getting the header from within the o-layout__main element instead of the first h2 in the document